### PR TITLE
[Mellanox] Add new cases for eeprom

### PR DESCRIPTION
**Pull request by JibinBao**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [x] 202012
- [x] 202205

#### What is the motivation for this PR?
Check the outputs are correct for "show interface transceiver eeprom -d", "sudo sfputil show eeprom -d".
1. Check if all expected keys exist in the outputs
2. When cable support dom, check the corresponding keys related to monitor exist, and the the corresponding value has the correct format

#### How did you do it?
Add new tests

#### How did you verify/test it?
Run script of test_check_sfp_eeprom.py

#### Any platform specific information?
Yes. Mellanox

#### Supported testbed topology if it's a new test case?
Any

### CODE REVIEW

**Comment by JibinBao** [comment]
> The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request. The detected issues may be old or new. For new issues, please try to fix them.
> 
> For old issues, it is not mandatory to fix them because they were not caused by this change. It is unfair to blame author of this pull request. But if you can take extra effort to fix the old issues as well, that would be great!
> 
> Detailed pre-commit check results: trim trailing whitespace.................................................Passedfix end of files.........................................................Passedcheck yaml...........................................(no files to check)Skippedcheck for added large files..............................................Passedcheck python ast.........................................................Passedflake8...................................................................Failed- hook id: flake8- exit code: 1tests/platform_tests/mellanox/test_check_sfp_eeprom.py:4:1: F401 'tests.common.fixtures.conn_graph_facts.conn_graph_facts' imported but unusedtests/platform_tests/mellanox/test_check_sfp_eeprom.py:16:64: F811 redefinition of unused 'conn_graph_facts' from line 4
> 
> To run the pre-commit checks locally, you can follow below steps:
> 
> 1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
>    the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
>    docker container.
> 2. Ensure that the `pre-commit` package is installed:
> 
> ```
> sudo pip install pre-commit
> ```
> 
> 3. Go to repository root folder
> 4. Install the pre-commit hooks:
> 
> ```
> pre-commit install
> ```
> 
> 5. Use pre-commit to check staged file:
> 
> ```
> pre-commit
> ```
> 
> 6. Alternatively, you can check committed files using:
> 
> ```
> pre-commit run --from-ref <commit_id> --to-ref <commit_id>
> ```
For this pre-commit issue, I am not sure how to resolve it. Maybe the pep8 check rule need to some modification? Because conn_graph_facts is a fixture Actually it has been used test case.

**Comment by JibinBao** [comment]
@nhe-NV

**Comment by liat-grozovik** [comment]
@Junchao-Mellanox could you please help to review?

File: tests/platform_tests/mellanox/util.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+def check_dom_monitor_key_and_data_format(expected_keys_and_pattern_dict, dom_monitor_data):
+    for key, pattern in expected_keys_and_pattern_dict.items():
+        assert key in dom_monitor_data, "Key {} doesn't exist in {}".format(key, dom_monitor_data)
+        assert re.match(pattern, dom_monitor_data[key]), \
+            "Value of {}:{} format is not correct. pattern is {}".format(
+                key, dom_monitor_data[key], pattern)
+
+
+def is_support_dom(duthost, port_index):
```
Comment (Junchao-Mellanox) [inline tests/platform_tests/mellanox/util.py:303]: Maybe it is better to use `ethtool -m sfp{index}`. Part of the data of "show interface transceiver eeprom -d" comes from `sfp.get_transceiver_bulk_status`. So, using `sfp.get_transceiver_bulk_status` to test "show interface transceiver eeprom -d" may not find real issues.

File: tests/platform_tests/mellanox/util.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    This method is to pares sfp eeprom infos, and return sfp_eeprom_info_dict
+    """
+    sfp_eeprom_info_dict = {}
+    for sfp_info in eeprom_infos.split("\n\n"):
+        intf, eeprom_info = parse_one_sfp_eeprom_info(sfp_info)
+        sfp_eeprom_info_dict[intf] = eeprom_info
+    return sfp_eeprom_info_dict
+
+
+def check_sfp_eeprom_info(duthost, sfp_eeprom_info, is_support_dom, show_eeprom_cmd):
```
Comment (Junchao-Mellanox) [inline tests/platform_tests/mellanox/util.py:184]: Could you please consider using compare the output of show eeprom command with ethtool/mlxlink output?

File: tests/platform_tests/mellanox/util.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+                key, dom_monitor_data[key], pattern)
+
+
+def is_support_dom(duthost, port_index):
+    """
+    This method is to check if cable support dom
+    """
+    bulk_status_str = get_transceiver_bulk_status(duthost, port_index)
+    bulk_status_str = bulk_status_str.replace('-inf', '\'-inf\'')
+    bulk_status_dict = eval(bulk_status_str)
```
Comment (Junchao-Mellanox) [inline tests/platform_tests/mellanox/util.py:309]: suggest use `literal_eval` instead of `eval`

**Comment by Junchao-Mellanox** [comment]
I am not sure the scope of this PR. A few general comments:

1. "show interface transceiver eeprom" does not cover everything, will you consider verify the value in STATE DB?
2. Will you consider platform API level verification? Like `sfp.get_transceiver_info`,  `sfp.get_transceiver_dom_threshold`...
3. Will you add case to verify EEPROM writing? For example, Y cable.

File: tests/platform_tests/mellanox/util.py (hunk 1 above)
Comment (keboliu) [reply tests/platform_tests/mellanox/util.py:303]: a specific field of the EEPROM can be different in the output of the "ethtool -m" and SONiC CLI output, no easy way to verify the SONIC CLI output with it.

File: tests/platform_tests/mellanox/util.py (hunk 2 above)
Comment (keboliu) [reply tests/platform_tests/mellanox/util.py:184]: for a specific field, the output from ethtool/mlxlink could have some differences with the SONiC CLI output, no easy way to directly compare of them.  only the Vendor OUI, Vendor Name, Vendor SN.. are the same and can be compared directly.

File: tests/platform_tests/mellanox/util.py (hunk 1 above)
Comment (keboliu) [reply tests/platform_tests/mellanox/util.py:303]: maybe can use mlxlink/ethtool to check whether relevant fields are valid.

File: tests/platform_tests/mellanox/util.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+            'VccLowAlarm': '3.1Volts',
+            'VccLowWarning': '3.135Volts'
+        }
+    }
+    """
+    one_sfp_eeprom_info_dict = {}
+    second_layer_dict = {}
+    previous_key = ""
+    top_key = ""
+    for line in sfp_eeprom_info.split("\n"):
```
[the remark this answers is not in the document]
Comment (JibinBao) [reply tests/platform_tests/mellanox/util.py:150]: Yes. for this case, case will raise failure

File: tests/platform_tests/mellanox/util.py (hunk 2 above)
Comment (JibinBao) [reply tests/platform_tests/mellanox/util.py:184]: Will use mlxlink tool to check if it supports dom.

File: tests/platform_tests/mellanox/util.py (hunk 3 above)
Comment (JibinBao) [reply tests/platform_tests/mellanox/util.py:309]: ok

**Comment by JibinBao** [comment]
> I am not sure about the scope of this PR. A few general comments:
> 
> 1. "show interface transceiver eeprom" does not cover everything, will you consider verify the value in STATE DB?
> 2. Will you consider platform API level verification? Like `sfp.get_transceiver_info`,  `sfp.get_transceiver_dom_threshold`...
> 3. Will you add case to verify EEPROM writing? For example, Y cable.

In the PR, we don't include these scope. If need, we can add one task to add them

**Comment by wangxin** [comment]
This PR added 2 new files, can you ensure that the new files can pass the pre-commit check? Thanks! @JibinBao

**Comment by JibinBao** [comment]
> This PR added 2 new files, can you ensure that the new files can pass the pre-commit check? Thanks! @JibinBao

Sure. There is only one issue in which fixture is used, but pep8 checks it is not imported. I will update it.

**Review by Junchao-Mellanox — DISMISSED** [review]

**Review by Junchao-Mellanox — APPROVED** [review]

**Pull request merged**